File: sdks/python/examples/run_roost_workflow.py
"""Minimal public API workflow for the Python SDK.

Required:
    OWLETTE_TOKEN
    OWLETTE_SITE_ID
    OWLETTE_ROOST_ID

Optional:
    OWLETTE_API_URL   default: https://owlette.app
    BUILD_DIR         default: ./dist
    OWLETTE_DEPLOY=1
"""

from __future__ import annotations

import asyncio
import os
import sys

from roost import DeployOptions, PushOptions, Roost, RoostApiError


def _progress(evt: object) -> None:
    phase = getattr(evt, "phase", "")
    if phase == "upload":
        print(f"upload {evt.uploaded}/{evt.total}")  # type: ignore[attr-defined]
    elif phase == "publish":
        print(f"publish attempt {evt.attempt}")  # type: ignore[attr-defined]


async def main() -> int:
    token = os.environ.get("OWLETTE_TOKEN")
    site_id = os.environ.get("OWLETTE_SITE_ID")
    roost_id = os.environ.get("OWLETTE_ROOST_ID")
    api_url = os.environ.get("OWLETTE_API_URL") or "https://owlette.app"
    build_dir = os.environ.get("BUILD_DIR", "./dist")
    deploy_enabled = os.environ.get("OWLETTE_DEPLOY") == "1"

    missing = [
        name
        for name, value in (
            ("OWLETTE_TOKEN", token),
            ("OWLETTE_SITE_ID", site_id),
            ("OWLETTE_ROOST_ID", roost_id),
        )
        if not value
    ]
    if missing:
        print("missing env: " + ", ".join(missing), file=sys.stderr)
        return 1

    assert token and site_id and roost_id

    async with Roost(token=token, api_url=api_url) as client:
        try:
            identity = await client.account.whoami()
            version = await client.account.version()
            print("identity", identity.email or identity.user_id or "api-key")
            print("api version", version.current)

            sites = await client.sites.list()
            print("sites", len(sites))
            for visible_site in sites[:10]:
                print("site", visible_site.id, visible_site.name)

            machines = await client.machines.list(site_id)
            print("machines", len(machines))
            for machine in machines[:10]:
                state = "online" if machine.online else "offline"
                print("machine", machine.id, machine.name, state)

            site = await client.sites.get(site_id)
            roost = await client.roosts.get(roost_id, site_id=site_id)
            print("site", site.id, site.name)
            print("roost", roost.roost_id, roost.name)

            result = await client.roosts.push(
                build_dir,
                roost_id,
                PushOptions(site_id=site_id, on_progress=_progress),
            )
            print(f"published v{result.version_number} {result.version_id}")

            if deploy_enabled:
                rollout = await client.roosts.deploy(
                    roost_id,
                    DeployOptions(site_id=site_id, version_id=result.version_id),
                )
                print("rollout", rollout.rollout_id, rollout.stage)

            return 0
        except RoostApiError as err:
            detail = err.problem.get("detail")
            print(f"api error {err.status} {err.code}: {detail}", file=sys.stderr)
            if err.request_id:
                print(f"request_id: {err.request_id}", file=sys.stderr)
            return 2


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))

File: sdks/node/examples/run-roost-workflow.ts
/**
 * Minimal public-api workflow:
 *   1. authenticate the token with /api/whoami
 *   2. inspect the target site and roost
 *   3. publish BUILD_DIR as a new roost version
 *
 * Required env vars:
 *   OWLETTE_TOKEN or ROOST_TOKEN
 *   OWLETTE_SITE_ID or ROOST_SITE_ID
 *   OWLETTE_ROOST_ID or ROOST_ID
 *   BUILD_DIR defaults to ./dist
 *
 * Optional:
 *   OWLETTE_API_URL or ROOST_BASE defaults to https://owlette.app
 *   OWLETTE_DEPLOY=1 queues a deploy after the publish
 */

import { Roost, RoostApiError } from '@owlette/sdk';

const token = process.env.OWLETTE_TOKEN ?? process.env.ROOST_TOKEN;
const apiUrl = process.env.OWLETTE_API_URL ?? process.env.ROOST_BASE ?? 'https://owlette.app';
const siteId = process.env.OWLETTE_SITE_ID ?? process.env.ROOST_SITE_ID;
const roostId = process.env.OWLETTE_ROOST_ID ?? process.env.ROOST_ID;
const buildDir = process.env.BUILD_DIR ?? './dist';
const shouldDeploy = process.env.OWLETTE_DEPLOY === '1';

for (const [name, value] of [
  ['OWLETTE_TOKEN or ROOST_TOKEN', token],
  ['OWLETTE_SITE_ID or ROOST_SITE_ID', siteId],
  ['OWLETTE_ROOST_ID or ROOST_ID', roostId],
]) {
  if (!value) {
    console.error(`missing env var: ${name}`);
    process.exit(1);
  }
}

const roost = new Roost({ token: token!, apiUrl });

try {
  const [identity, version] = await Promise.all([
    roost.account.whoami(),
    roost.account.version(),
  ]);
  console.log('api', version.current, 'user', identity.email ?? identity.userId);
  console.log('key', identity.key?.keyPrefix ?? 'session', 'primary site', identity.primarySiteId);

  const [site, currentRoost] = await Promise.all([
    roost.sites.get(siteId!),
    roost.roosts.get(roostId!, { siteId: siteId! }),
  ]);
  console.log('site', site.id, site.name);
  console.log('roost', currentRoost.roostId, currentRoost.name);

  const published = await roost.roosts.push(buildDir, roostId!, {
    siteId: siteId!,
    description: `node sdk publish ${new Date().toISOString()}`,
    onProgress: (evt) => {
      if (evt.phase === 'upload') console.log('upload', `${evt.uploaded}/${evt.total}`);
      if (evt.phase === 'publish') console.log('publish attempt', evt.attempt);
    },
  });

  console.log('published', `v${published.versionNumber}`, published.versionId);

  if (shouldDeploy) {
    const deploy = await roost.roosts.deploy(roostId!, {
      siteId: siteId!,
      versionId: published.versionId,
    });
    console.log('deploy queued', deploy.rolloutId, deploy.stage);
  }
} catch (err) {
  if (err instanceof RoostApiError) {
    console.error('api error', err.status, err.code, err.problem.detail ?? err.message);
  } else {
    console.error('unexpected error', err);
  }
  process.exit(1);
}
